### PR TITLE
/send command; ability to add bots to group chats

### PR DESCRIPTION
**Pull request by alwx**
Done.
Includes several important fixes for `parameter-boxes` and other components.

### CODE REVIEW

File: src/status_im/chat/handlers/input.cljs [hunk 1]
```diff
@@ -332,3 +337,47 @@
...
     (let [chat-id (or chat-id current-chat-id)]
       (assoc-in db [:chats chat-id :seq-argument-input-text] text))))
+
+(handlers/register-handler
+  :update-text-selection
+  (handlers/side-effect!
+    (fn [{:keys [current-chat-id] :as db} [_ selection]]
+      (let [input-text (get-in db [:chats current-chat-id :input-text])
+            command    (input-model/selected-chat-command db current-chat-id input-text)]
+        (when (and (= selection (+ 2 (count (get-in command [:command :name]))))
```
Comment (rasom) [inline src/status_im/chat/handlers/input.cljs]: @alwx it's the code is not super clear here, could you add some comment? I mean this `+ 2` needs some explanation, `(when what-exactly? ...)`

File: src/status_im/chat/handlers/input.cljs [hunk 2]
```diff
@@ -332,3 +337,47 @@
...
+          (dispatch [:chat-input-focus :seq-input-ref]))
+        (dispatch [:set-chat-ui-props {:selection selection}])
+        (dispatch [:load-chat-parameter-box (:command command)])))))
+
+(handlers/register-handler
+  :select-prev-argument
+  (handlers/side-effect!
+    (fn [{:keys [chat-ui-props current-chat-id] :as db} _]
+      (let [arg-pos (input-model/argument-position db current-chat-id)]
+        (when (> arg-pos 0)
```
Comment (rasom) [inline src/status_im/chat/handlers/input.cljs]: `pos?`

File: src/status_im/chat/handlers/input.cljs [hunk 3]
```diff
@@ -332,3 +337,47 @@
...
+          (let [input-text (get-in db [:chats current-chat-id :input-text])
+                new-sel    (->> (input-model/split-command-args input-text)
+                                (take (+ 1 arg-pos))
```
Comment (rasom) [inline src/status_im/chat/handlers/input.cljs]: `inc`

File: src/status_im/chat/handlers/input.cljs [hunk 4]
```diff
@@ -332,3 +337,47 @@
...
+
+(handlers/register-handler
+  :select-next-argument
+  (handlers/side-effect!
+    (fn [{:keys [chat-ui-props current-chat-id] :as db} _]
+      (let [arg-pos (input-model/argument-position db current-chat-id)]
+        (let [input-text   (get-in db [:chats current-chat-id :input-text])
+              command-args (cond-> (input-model/split-command-args input-text)
+                                   (input-model/text-ends-with-space? input-text) (conj ""))
+              _ (log/debug "ALWX command-args" input-text command-args arg-pos)
```
Comment (rasom) [inline src/status_im/chat/handlers/input.cljs]: debug message

File: src/status_im/chat/handlers/input.cljs [hunk 5]
```diff
@@ -332,3 +337,47 @@
...
+              new-sel      (->> command-args
+                                (take (+ 3 arg-pos))
```
Comment (rasom) [inline src/status_im/chat/handlers/input.cljs]: Why `+ 3`?

File: src/status_im/chat/models/input.cljs [hunk 6]
```diff
@@ -81,26 +81,28 @@
...
+      (count seq-arguments)
+      (let [subs-input-text (subs input-text 0 selection)]
+        (if subs-input-text
+          (let [args             (split-command-args subs-input-text)
+                argument-index   (dec (count args))
+                ends-with-space? (text-ends-with-space? subs-input-text)]
+            (if ends-with-space?
+              argument-index
+              (dec argument-index)))
+          -1)))
```
Comment (rasom) [inline src/status_im/chat/models/input.cljs]: Magic value `-1`, better to have some constant here.

**Comment by flexsurfer** [comment]
image

**Comment by alwx** [comment]
It's finally done.
It was problematic to make commands work in group chats, but it is also implemented. However, there are things that can be implemented, but it's better to postpone them — for instance, if we want to execute `/send` command in 1-to-1 chat without any contact selection, we need to modify the API. I see two options:
1) We can introduce scopes and allow commands to be `group-only` or `1-to-1-only`. By doing this, we will allow users to have different commands with the same name;
2) We can transform command's `parameters` to be functions, and there we will be able to decide how many parameters to return.

Both changes are breaking, and it's worth doing something only after the hackathon.

**Comment by jarradh** [comment]
nice work and agreed on both points - they can wait til after hackathon

**Comment by annadanchenko** [comment]
@alwx Can't send 1.1234 ETH - no Confirm transaction screen is shown

in 1-1 chat send 1.12345 ETH -> no screen to confirm transaction is shown, app freezes. iPhone 7 plus, iOS 10.3.1
[nosend.log.txt](https://github.com/status-im/status-react/files/1038971/nosend.log.txt)
https://drive.google.com/open?id=0Bz3t9zSg1wb7Wno0Wnllc1pEWjQ

**Comment by annadanchenko** [comment]
On Android: password input shows plain text instead of dots:
screenshot_20170530-173300

**Comment by annadanchenko** [comment]
on both iOS and Android: can send /request with not-number, e.g `/request abc`
it's send as Requesting NaN ETH. 
img_0101

**Comment by annadanchenko** [comment]
both iOS and Android: can `/send non_existing_contact 123` and get error on command handling. for this use /send comand but do not select contact from the list, just type some random word and sum and tap send icon
screenshot_20170530-221912
screenshot_20170530-221907

**Comment by annadanchenko** [comment]
if recipient name contains double quote " then it's not possible to send eth:
1. name has " at the end: send here on ios" -> name is shown again for selection after sum is entered. tap on send icon does nothing
screenshot_20170530-224215


2. name has " inside it and at the end: send "here on ios" -> no screen Send transaction 
screenshot_20170530-224247

**Comment by alwx** [comment]
Everything (except https://github.com/status-im/status-react/pull/1227#issuecomment-304930134, because it has been moved to another task) has been fixed. 
@annadanchenko, please, check it.

**Comment by flexsurfer** [comment]
On Android: password input shows plain text instead of dots: - still can be reproduced

**Comment by alwx** [comment]
Sorry, I've committed the fix.

**Comment by alwx** [comment]
So now it should be checked and probably merged :)

**Comment by annadanchenko** [comment]
/send Jarrad aaa -> Error on command handling
img_1553




on last build Branch: feature/send-command
Android: http://artifacts.status.im:8081/artifactory/pull-requests/im.status.ethereum-a33832.apk
iOS: https://i.diawi.com/Ytb2EB

**Comment by annadanchenko** [comment]
in old `/send` if it’s not enough ether then we show error “Insufficient funds…” when user tap on “send” and do not show transaction to sign. In new `/send`, we make user to confirm transaction and only than show error on Insufficient funds. Expected: works as in old `/send

**Comment by alwx** [comment]
I've fixed both issues.

File: src/status_im/commands/handlers/loading.cljs [hunk 7]
```diff
@@ -16,26 +16,22 @@
...
-      (dispatch [::fetch-commands! contact])))
-
+  [{:keys [current-chat-id contacts chats]} [jail-id callback]]
+  (let [identity    (or jail-id current-chat-id)
+        contact-ids (if (get contacts identity)
+                      [identity]
+                      (->> (get-in chats [identity :contacts])
+                           (map :identity)
+                           (into [])))]
+    (when (seq contacts)
```
Comment (flexsurfer) [inline src/status_im/commands/handlers/loading.cljs]: what does this check do?

**Comment by annadanchenko** [comment]
In Group chat: if contact A sends to B some ether then there are 2 messages shown:
1. Contact A, empty line, sum
2. Contact A, empty line, NaN ETH
I would expect 1 message with info about who sends ether; to whom ether is sent, sum sent
screenshot_20170602-130958

File: src/status_im/commands/handlers/loading.cljs (hunk 7 above)
Comment (alwx) [inline src/status_im/commands/handlers/loading.cljs]: what do you mean?
```
cljs.user=> (seq nil)
nil
cljs.user=> (seq [])
nil
cljs.user=> (seq [1 2 3])
(1 2 3)
```

Comment (rasom) [inline src/status_im/commands/handlers/loading.cljs]: https://github.com/bbatsov/clojure-style-guide#nil-punning

Comment (flexsurfer) [inline src/status_im/commands/handlers/loading.cljs]: thank you, It's not obvious without this link

**Comment by annadanchenko** [comment]
In 1-1 chat, Contact A requesting 0.32 ETH. Contact B taps on the request message in chat but input field contains "/send 0.32 09ef186e17b87bc420fe84f4d74d327dpa848af11" and if try to send it then error "Amount is not a valid number" shown. 
"09ef186e17b87bc420fe84f4d74d327dpa848af11" is an address of Contact B that is strange as ether is to be sent to Contact A.
Expected: when user taps on request message then all data is filled in, so it's ok to use Send
img_1554

**Comment by alwx** [comment]
@annadanchenko done. 
I've updated both /request and /send commands, checked the communication between two devices (Android and iOS) and updated the code.

It can be tested now.

**Comment by rasom** [comment]
That's how `/send` and `/receive` are shown in group chat. 
Both send/request were sent to Jarrad's account.

**Comment by rasom** [comment]
@alwx please resolve conflicts

**Comment by annadanchenko** [comment]
@alwx "Recipient address must be specified" error if tap on Request and try to send valid command. Issue happens only if in the current session in 1-1 chat there were no /send commands send before this. 
video: https://drive.google.com/open?id=0Bz3t9zSg1wb7T1ZuU3BsYWx5X0E
Steps: 
1. Launch app
2. Receive request for 0.123 ETH
3. Tap on request. As a result command `/send "iphone6 send" 0.123` is shown
4. Tap on Send button

Currently: error "Recipient address must be specified" is shown
Expected: Transaction screen to confirm transaction

**Comment by annadanchenko** [comment]
@alwx `[missing {{amount}} value] ETH` is shown in group chat to all members (except sender)
screenshot_20170628-134148



1. create group chat
2. As contact A send to contact B 0.23 ETH
3. As contact B open group chat and check messages

Expected: message like X ETH is sent from A to B
Currently: there are 2 messages shown to recipient and other group members:
1. Contact A 0.23 ETH
2. [missing {{amount}} value] ETH - this message is wrong

File: bots/wallet/bot.js [hunk 8]
```diff
@@ -1,13 +1,280 @@
...
+
+status.defineSubscription(
+    "calculatedFee",
+    {value: ["sliderValue"], tx: ["transaction"]},
+    function (params) {
+        return calculateFee(params.value, params.tx);
+    }
+);
+
+function getFeeExplaination(n) {
```
Comment (flexsurfer) [inline bots/wallet/bot.js]: Explanation

File: bots/wallet/bot.js [hunk 9]
```diff
@@ -1,13 +1,280 @@
...
+                                paddingRight: 16,
+                                paddingBottom: 20
+                            }
+                        },
+                        I18n.t('send_specify_amount')
+                    ),
+                    status.components.touchable(
+                        {
+                            onPress: status.components.dispatch([status.events.FOCUS_INPUT, []])
+                        },
```
Comment (flexsurfer) [inline bots/wallet/bot.js]: difficult to read code with styles maybe better to move styles in the separate file like in clojure we do?

File: src/status_im/chat/views/choosers/choose_contact.cljs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    (dispatch [:select-next-argument])))
+
+(defn render-row [arg-index bot-db-key]
+  (fn [contact _ _]
+    (list-item
+      ^{:key contact}
+      [contact-view {:contact  contact
+                     :on-press #(select-contact arg-index bot-db-key contact)}])))
+
+(defn choose-contact-view [{arg-index :index}]
```
Comment (flexsurfer) [inline src/status_im/chat/views/choosers/choose_contact.cljs]: just interesting why don't you use defview macro, not only here

File: src/status_im/chat/handlers/commands.cljs [hunk 11]
```diff
@@ -6,15 +6,27 @@
...
   (handlers/side-effect!
-    (fn [{:keys [contacts current-account-id] :as db}
+    (fn [{:keys [contacts current-account-id chats] :as db}
          [_ {{:keys [command params content-command type]} :content
-             :keys [message-id chat-id on-requested jail-id] :as message} data-type]]
-      (let [jail-id (or jail-id chat-id)]
-        (if-not (get-in contacts [jail-id :commands-loaded])
+             :keys [message-id from chat-id on-requested jail-id] :as message} data-type]]
+      (let [jail-id (or jail-id chat-id)
+            jail-id (if (get-in chats [jail-id :group-chat])
```
Comment (flexsurfer) [inline src/status_im/chat/handlers/commands.cljs]: i saw in code using of `'` for redefine, i think it's good practice `jail-id'`

**Comment by rasom** [comment]
1. create acc using `develop` branch
2. use `/sent`/`/receive` commands in 1-1 chat
3. upgrade app using this branch
4. open 1-1 chat with `/send`/`/receive` messages
5. see crash

```
Jul  7 20:02:27 StatusIm[3887] <Notice>: Error rendering component (in Uz > reagent1 > Vz > reagent38 > Lz > reagent45 > Pf > ListView >  > bz > reagent46 > n > Mf > message-content > wz > az > nz)
Jul  7 20:02:27 StatusIm[3887] <Notice>: null is not an object (evaluating 'l.a')
Jul  7 20:02:27 StatusIm[3887] <Notice>: Unhandled JS Exception: null is not an object (evaluating 'l.a')
```

**Comment by rasom** [comment]
1. That's how receive/send messages look at first run


2. That's how them look after restart (after 2-3 messages look fine)

**Pull request merged**